### ISSUE TITLE
Can not write to /tmp when securityContext.readOnlyRootFilesystem set to true

### ISSUE THREAD

**Issue by JulesMac** [knative/specs#98]
Looking at the [spec](https://github.com/knative/specs/blob/main/specs/serving/runtime-contract.md#default-filesystems) it seems that /tmp should be RW but this does not seem to be the case. My test app simply tried to write a file to /tmp - it work when readOnlyRootFilesystem is false (and fails when true)

```
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: test
  namespace: testns
spec:
  template:
    spec:
      containers:
      - image: "localhost:5000/test:1.0"
        securityContext:
          runAsNonRoot: true
          readOnlyRootFilesystem: true
```

**Comment by dprotaso** [comment]
sorry clarify are you using Knative OSS or a vendor distribution ie. Google's CloudRun

**Comment by JulesMac** [comment]
Setup locally on my laptop (docker desktop) but also in azure using these instructions
https://knative.dev/docs/install/yaml-install/serving/install-serving-with-yaml/

**Comment by dprotaso** [comment]
Thanks for the clarification. 

The relevant two clauses from the spec are

https://github.com/knative/specs/blob/main/specs/serving/runtime-contract.md#root
> Additionally, users may set the readOnlyRootfilesystem field in the SecurityContext to request a read-only filesystem

https://github.com/knative/specs/blob/main/specs/serving/runtime-contract.md#default-filesystems
> In addition to the filesystems recommended in the OCI, the following filesystems [MUST](https://github.com/knative/serving/blob/main/test/conformance/runtime/filesystem_test.go) be provided:
> /tmp MUST be Read-write.

I could see this going either way where
1) `/tmp` falls under the scope `readOnlyRootfilesystem: true` and MUST be readonly
2) `/tmp` is an exception to `readOnlyRootfilesystem: true` and MUST always remain RW

My recommendation is for 1) with the caveat that we eventually allow for  `emptyDir` volume support. Thus if users want a readonly FS and a RW /tmp dir then can mount an emptyDir.

Going to assign the to the KTC to discss

/assign @knative/trademark-committee

**Comment by dprotaso** [comment]
> issue I have is that there is an explicit check to stop me mounting emptydir (by enabling [this](https://knative.dev/docs/serving/configuration/feature-flags/#kubernetes-emptydir-volume) feature flag) at /tmp - because of [this](https://github.com/knative/serving/blob/main/pkg/apis/serving/k8s_validation.go#L47) I imagine.

Yeah - this is in the OSS implementation. In other distributions (ie. that don't use the OSS code) you won't have that workaround since volume support is not part of the specification.

**Comment by JulesMac** [comment]
I’ve just re-read this

https://github.com/knative/specs/blob/main/specs/serving/runtime-contract.md#root

it actually explicitly says 

Containers MUST use the provided temporary storage areas (see [Default Filesystems](https://github.com/knative/specs/blob/main/specs/serving/runtime-contract.md#default-filesystems)) for working files and caches.

so seems that  /tmp is required to be RW?

**Comment by evankanderson** [comment]
Ugh, this is an annoying one.

When I wrote this, I'll admit that I was thinking first about Google's Cloud Run implementation and the underlying bind-mount mechanism in Linux, without specifically looking at how that was exposed through Kubernetes. IMO, the ideal solution would allow an overlay mount (`tmpfs` or simply a different temp area) on the `/tmp` mountpont without disrupting the existing volume structure under that point. Unfortunately, Kubernetes volume mounts don't provide the "union mount" option, so we're stuck with:

1. Automatically add an `emptyDir` mount at `/tmp`, possibly overwriting existing `/tmp` contents in a volume.
2. Ask users to explicitly add an `emptyDir` mount on Knative OSS for `/tmp` if they set `readOnlyRootfilesystem`

I sort of want to do 1, but experience with a similar strategy for `/var/log` makes me hesitant. This is absolutely a place where Kubernetes makes it hard to do the right thing, and we should strive to be better, but I'm not quite sure what that means.

**Comment by dprotaso** [comment]
My comment below is in regards to the OSS code

re: 1) We did that with `/var/log` in the past and it ended up breaking some images (ie. nginx) - so there might be a potential issue there. We currently only add that the `/var/log` volume when that [feature is enabled](https://github.com/knative/serving/blob/08e09bc9547af527c3a466635997a91c044e407a/config/core/configmaps/observability.yaml#L43-L46). We could do something similar for `/tmp` when `readOnlyRootfilesystem: true` - what would minimize the impact. I agree that if K8s supported an RW overlay at a specific mount point that would make this much easier.

I want us (and the KTC) to reason about my question here: https://github.com/knative/specs/issues/98#issuecomment-1127934093 and answer whether we consider `/tmp` to be part of the root filesystem and thus `readOnlyRootfilesystem: true` behaves as it should?

**Comment by nitusharma2022** [comment]
any solution for this, i am struggling for the same

**Comment by dprotaso** [comment]
This was fixed Knative Serving v1.13

https://github.com/knative/serving/pull/14719
